### PR TITLE
Linear connector

### PR DESCRIPTION
**Pull request by Weves**
Screenshot 2023-08-17 at 2 55 43 PM
Screenshot 2023-08-17 at 2 55 39 PM
Screenshot 2023-08-17 at 2 55 18 PM

### CODE REVIEW

File: backend/danswer/connectors/factory.py [hunk 1]
```diff
@@ -22,8 +23,6 @@
 from danswer.connectors.web.connector import WebConnector
 from danswer.connectors.zulip.connector import ZulipConnector
 
-_NUM_SECONDS_IN_DAY = 86400
```
Comment (yuhongsun96) [inline backend/danswer/connectors/factory.py:25]: If ever we need this, it should be in the constants file

File: backend/danswer/connectors/linear/connector.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+from danswer.connectors.interfaces import LoadConnector
+from danswer.connectors.interfaces import PollConnector
+from danswer.connectors.interfaces import SecondsSinceUnixEpoch
+from danswer.connectors.models import Document
+from danswer.connectors.models import Section
+from danswer.utils.logger import setup_logger
+
+logger = setup_logger()
+
+_NUM_RETRIES = 5
```
Comment (yuhongsun96) [inline backend/danswer/connectors/linear/connector.py:22]: shared connector per doc retries config?

File: backend/danswer/connectors/linear/connector.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    headers = {
+        "Authorization": api_key,
+        "Content-Type": "application/json",
+    }
+
+    response: requests.Response | None = None
+    for i in range(_NUM_RETRIES):
+        try:
+            response = requests.post(
+                "https://api.linear.app/graphql",
```
Comment (yuhongsun96) [inline backend/danswer/connectors/linear/connector.py]: nit: global "static" on top of file?

File: backend/danswer/connectors/linear/connector.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+                headers=headers,
+                json=request_body,
+                timeout=60,
```
Comment (yuhongsun96) [inline backend/danswer/connectors/linear/connector.py]: Probably can be shared across modules too, connector connection timeout or something

File: backend/danswer/connectors/linear/connector.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+                )
+
+            return response
+        except Exception as e:
+            if i == _NUM_RETRIES - 1:
+                raise e
+
+            logger.warning(f"A Linear GraphQL error occurred: {e}. Retrying...")
+
+    raise RuntimeError(
```
Comment (yuhongsun96) [inline backend/danswer/connectors/linear/connector.py:54]: why is this here?

Comment (yuhongsun96) [reply backend/danswer/connectors/linear/connector.py:54]: mypy reasons?

File: backend/danswer/connectors/linear/connector.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+
+class LinearConnector(LoadConnector, PollConnector):
+    def __init__(
+        self,
+        batch_size: int = INDEX_BATCH_SIZE,
+    ) -> None:
+        self.batch_size = batch_size
+
+    def load_credentials(self, credentials: dict[str, Any]) -> dict[str, Any] | None:
+        self.linear_api_key = credentials["linear_api_key"]
```
Comment (yuhongsun96) [inline backend/danswer/connectors/linear/connector.py]: Should this be initialized to None in init?

File: backend/danswer/connectors/linear/connector.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    def poll_source(
+        self, start: SecondsSinceUnixEpoch, end: SecondsSinceUnixEpoch
+    ) -> GenerateDocumentsOutput:
+        start_time = datetime.fromtimestamp(start, tz=timezone.utc)
+        end_time = datetime.fromtimestamp(end, tz=timezone.utc)
+
+        yield from self._process_issues(start_str=start_time, end_str=end_time)
+
+
+if __name__ == "__main__":
```
Comment (yuhongsun96) [inline backend/danswer/connectors/linear/connector.py:213]: remove this junk!

**Review by yuhongsun96 — APPROVED** [review]
nit comments, ideally address but approving.

For future, maybe connector PRs should include screencap of frontend so we can see what it looks like without needing to run the code?

File: backend/danswer/connectors/linear/connector.py (hunk 5 above)
Comment (Weves) [reply backend/danswer/connectors/linear/connector.py:54]: mypy + to let us know loudly if this ever happens

File: backend/danswer/connectors/linear/connector.py (hunk 3 above)
Comment (Weves) [reply backend/danswer/connectors/linear/connector.py]: done!

File: backend/danswer/connectors/linear/connector.py (hunk 2 above)
Comment (Weves) [reply backend/danswer/connectors/linear/connector.py:22]: Hmm, I'm personally more in-favor of allowing each connector to set this, since it depends on the reliability of the API

File: backend/danswer/connectors/linear/connector.py (hunk 4 above)
Comment (Weves) [reply backend/danswer/connectors/linear/connector.py]: Same comment as for retries, but I don't have too strong of an opinion

File: backend/danswer/connectors/linear/connector.py (hunk 6 above)
Comment (Weves) [reply backend/danswer/connectors/linear/connector.py]: Done!

File: backend/danswer/connectors/linear/connector.py (hunk 7 above)
Comment (Weves) [reply backend/danswer/connectors/linear/connector.py:213]: after discussing, keeping in to allow for easier manual testing

Comment (Weves) [reply backend/danswer/connectors/linear/connector.py:213]: future connectors should have something like this + it should be added to all existing connectors

**Comment by ericmigi** [comment]
Does it pull in all past issues during the initial sync?

**Comment by Weves** [comment]
@ericmigi the initial run should pull in _all_ issues! And then, after that, we will only pull in issues changed in the last X minutes (where X is the time since the start of the last successful run plus an offset).

Hopefully that answers your question! If not let me know, and I'm happy to dig in deeper.
